BREAKING: Android: Correct value of Dimensions.get('screen').fontScale

rigdern opened the pull request:
The PR description has been updated to reflect the new approach.

**Breaking Change Summary**

On Android, the following properties now return a different number:
  - `Dimensions.get('window').fontScale`
  - `Dimensions.get('screen').fontScale`
  - `PixelRatio.getFontScale()`

This is a breaking change to anyone who was using these properties because the meaning of these properties has now changed.

These properties used to return a value representing font scale times density ([`DisplayMetrics.scaledDensity`](https://developer.android.com/reference/android/util/DisplayMetrics.html#scaledDensity)). Now they return a value representing just font scale ([`Configuration.fontScale`](https://developer.android.com/reference/android/content/res/Configuration.html#fontScale)).

**PR Description**

This PR changes a few things:
  - Correctly exposes the font scale to JavaScript as `Dimensions.get('screen').fontScale`. UIManager was exporting `DisplayMetrics.scaledDensity` under the name `fontScale`. However, this isn't correct because `scaledDensity` is actually `density * fontScale`. We now correctly export just `fontScale` under the name `fontScale`.
  - `didUpdateDimensions` now fires whenever `fontScale` changes.
  - `Dimensions` now fires a `change` event whenever a property within the `Dimensions` object changes. The primary benefit of this event over `didUpdateDimensions` is that this event provides an event object which has the same shape on iOS and Android. The event object for `didUpdateDimensions` uses different properties names on iOS than it does on Android.

**Test plan (required)**

Verified that the API works for different font settings in a test app. Also, verified that the `didUpdateDimensions` and `change` events fire with the proper information when changing the font settings and when rotating the device.

Adam Comella
Microsoft Corp.

mkonicek commented:
Since it looks like the method and event deal purely with fonts, what do you think of the following naming?

`getContentSizeMultiplier` ->`getFontScale`
`didUpdateContentSizeMultiplier` -> `didUpdateFontScale`

But I see on iOS this reads `bridge.accessibilityManager.multiplier` so maybe it's related to other things than fonts?

Should the method and event be declared and documented in a JS file somewhere?

astreet commented:
1) Just fyi, the activity will not restart if you've set `android:configChanges="fontScale"`. I don't think that's very common though
2) If we did have a listener, though, we could just send this down in the initial UIManager constants instead of exposing a getter.

rigdern commented:
@mkonicek

> Should the method and event be declared and documented in a JS file somewhere?

Functionality-wise, everything works fine without declaring the method or event in a JS file. [`UIManager.js`](https://github.com/facebook/react-native/blob/3aa25f79172159e3079568d0131623054916fb59/Libraries/ReactNative/UIManager.js) adds most of its methods automatically based on what was exported from native.

You raise a good point about documentation. Do you have any suggestions of where to document it? It doesn't look like there's any API reference page for `UIManager` on the website. There are at least a couple of other `UIManager` methods that aren't well documented:
  - [`sendAccessibilityEvent`](https://github.com/facebook/react-native/blob/3aa25f79172159e3079568d0131623054916fb59/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java#L527-L530) is documented with [a snippet on the website](http://facebook.github.io/react-native/releases/0.37/docs/accessibility.html#sending-accessibility-events-android) but doesn't have any API reference.
  - [`takeSnapshot`](https://github.com/facebook/react-native/blob/3aa25f79172159e3079568d0131623054916fb59/React/Modules/RCTUIManager.m#L1383-L1387) appears in a UIExplorer example but isn't mentioned on the website.

Perhaps the same solution should be used for documenting all of these UIManager methods.

> Since it looks like the method and event deal purely with fonts, what do you think of the following naming?
>
>  `getContentSizeMultiplier` ->`getFontScale`
>  `didUpdateContentSizeMultiplier` -> `didUpdateFontScale`
>
> But I see on iOS this reads bridge.accessibilityManager.multiplier so maybe it's related to other things than fonts and it makes sense to keep the name more generic?

I don't have a strong preference here.

I'm not an expert on the subject but looking at the iOS and Android docs, the native sources of `getContentSizeMultiplier` appear to be about font scaling:
  - Android's [`fontScale`](https://developer.android.com/reference/android/content/res/Configuration.html#fontScale) property
  - iOS's [`preferredContentSizeCategory`](https://developer.apple.com/reference/uikit/uiapplication/1623048-preferredcontentsizecategory?language=objc) property

Since this appears to be about font scaling, I can see why the name `getFontScale` makes sense. Here are some reasons `getContentSizeMultiplier` makes sense:
  - Font scaling happens automatically by RN so the reason an app developer would want to get at this value would be to manually scale their content (e.g. view sizes) to look better with the automatically scaled text.
  - iOS picked the name `preferredContentSizeCategory` which is named after content size rather than font scaling (although the docs indicate it's about font scaling).

If you want me to rename it to `getFontScale`, let me know. Note that my PR for `getContentSizeMultiplier` was already merged on iOS. Is it the case that we can rename it as long as it hasn't yet made it into any releases? What about releases vs RCs?

rigdern commented:
@astreet 

> 1. Just fyi, the activity will not restart if you've set `android:configChanges="fontScale"`. I don't think that's very common though
> 2. If we did have a listener, though, we could just send this down in the initial UIManager constants instead of exposing a getter.

Do you propose we do anything for the `android:configChanges="fontScale"` case? You mentioned it's uncommon and my team doesn't use that configuration so I'm happy to ignore the case until somebody needs it.

For (2), do you propose we expose that value to JS through `Dimensions` like RN does for width and height?

As I mentioned to Martin, I discovered that RN Android already exposes a `fontScale` property off of `Dimensions`. Do you know what that `fontScale` property is for and how it relates to the value in the `getContentSizeMultiplier` API I'm adding?

> Confusingly, RN Android already exposes [a `fontScale` property off of `Dimensions`](https://github.com/facebook/react-native/blob/3aa25f79172159e3079568d0131623054916fb59/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L111-L112). It's based on Android's [`scaledDensity`](https://developer.android.com/reference/android/util/DisplayMetrics.html#scaledDensity) property. Reading the docs, I'm not entirely sure what it means. In testing, it was giving a different value than `getContentSizeMultiplier` and `getContentSizeMultiplier`'s result was preferable.

astreet commented:
Whoops, I think you found a bug/misnaming in RN: According to http://androidxref.com/4.2_r1/xref/packages/apps/Settings/src/com/android/settings/Display.java#99, `scaledDensity` which we expose here https://github.com/facebook/react-native/blob/abb8ea3aea686e2cd881a61fbc66d137857ef422/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L111 as `fontScale`, is actually `fontScale * density`. But it seems to be exposed and used as an actual fontScale in our JS code? I'm pretty confused, this seems very clowny as it would indicate some things using this weren't very well tested :P @foghina, am I reading this wrong?

mkonicek commented:
> You raise a good point about documentation. Do you have any suggestions of where to document it? It doesn't look like there's any API reference page for UIManager on the website. There are at least a couple of other UIManager methods that aren't well documented:

I agree, I checked the website before asking the original question and didn't see any docs for `UIManager` :( http://facebook.github.io/react-native/docs/getting-started.html 

We do have a `UIManager.js` though, would it make sense to add the function there and document it? https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/UIManager.js

> Here are some reasons getContentSizeMultiplier makes sense:

Good points, let's keep `getContentSizeMultiplier`.

> Confusingly, RN Android already exposes a fontScale property off of Dimensions.

Thanks for noticing this! It might be wrong and maybe we should deprecate it as part of this PR?
To make it more confusing we also have [`PixelRatio.getFontScale`](https://github.com/facebook/react-native/blob/master/Libraries/Utilities/PixelRatio.js#L67). Git blame suggests @foghina, I pinged him on messenger asking to comment here.

foghina commented:
Ok. So.

1. I don't think we should add an event for when the font scale changes. Using `android:configChanges="fontScale"` is untested and I have a hunch it wouldn't really work, since we'd need to do a layout pass when it happens.
2. As @astreet pointed out, `scaledDensity = density * fontScale`. The reason why I exposed `scaledDensity` is because this is what the system uses when calculating the pixel value of an `sp` unit. To demonstrate, I've created a sample app with a `TextView` with `textSize="18sp"`, then got the pixel value via `TextView#getTextSize()`. As you can see, the pixel value is always `18 * scaledDensity`: http://imgur.com/a/DJtnG. I think the initial use case for this constant was to manually calculate pixel values, so `scaledDensity` made sense. Not sure where `fontScale` would be useful.

As for `getContentSizeMultiplier`, it's not clear to me what the use cases for it are. Sounds like you can easily calculate the `fontScale` by using the density and scaledDensity, which you can get from `PixelRatio` / `Dimensions`. IMO iOS should also expose it the same way, we should refrain from adding random getters on `UIManagerModule`.

astreet commented:
Ok, thanks Felix. I don't think we should be doing anything new then, except for maybe renaming fontScale to something more accurate (this would be breaking though) or exposing a convenience method to PixelRatio.

rigdern commented:
@foghina @astreet 

You've convinced me that we don't need to add `getContentSizeMultiplier` to `UIManager` because Android already has APIs for extracting this value. Should we revert eddc2c90f693739273583b74fc70148182b9f143 which added this API to iOS?

As Felix mentioned, on Android we can calculate `fontScale` as `scaledDensity / density`. Our use case for `fontScale` is to adjust how things are sized to match scaled fonts. RN's size props (e.g. `width`, `padding`) assume the unit is `dp`. We multiply a size by `fontScale` in JS so that it ends up being treated as though its unit was `sp` by native. What is a use case for `scaledDensity` in JS? I imagine this is only useful if RN native code won't be converting from `dp` to `pixels` for you.

For iOS, we could expose `density` and `scaledDensity` just like RN Android does. Does anyone know what iOS APIs give the equivalent of Android's `density` and `scaledDensity`?

@foghina, you mentioned you don't think we should add random getters to UIManager. In general, should apps be directly calling any UIManager APIs from JS? If not:
  - What's the reasoning?
  - [Earlier in this issue](https://github.com/facebook/react-native/pull/11008#issuecomment-262427766), I mentioned to Martin that UIManager has some APIs that aren't well documented (e.g. `sendAccessibilityEvent` and `takeSnapshot`). We were thinking about how to document them better. Instead, maybe they should be exposed to JS elsewhere.

astreet commented:
Yeah, I would say we should revert eddc2c9. I'm not sure about the iOS APIs -- do they even have system-wide font scaling?

It also looks to me like our one internal use of `getFontScale` is incorrect: `const adaptiveFontSize = Math.min(12 * PixelRatio.getFontScale(), 22);` seems to not realize that getFontScale also includes density (which will be applied on the native side already)... and really that's not even correct because we will also apply the scaling factor automatically on the native side. I think we should actually just change the `fontScale` that's exported from UIManager to be the actual font scale and not the scaled density. I agree that scaled density doesn't seem useful.

rigdern commented:
> I'm not sure about the iOS APIs -- do they even have system-wide font scaling?

Yes, they do. eddc2c90f693739273583b74fc70148182b9f143 exposed the current font scale to JS.

---

Currently, Android exposes `DisplayMetrics.density` as `scale` and `DisplayMetrics.scaledDensity` as `fontScale`.

@astreet @foghina What do you think of this proposal:
  - On Android:
    - Modify `UIManager` so that `fontScale` is `DisplayMetrics.scaledDensity / DisplayMetrics.density` instead of just `DisplayMetrics.scaledDensity`.
  - On iOS:
    - Remove `getContentSizeMultiplier ` from iOS by reverting eddc2c90f693739273583b74fc70148182b9f143
    - Expose a `UIManager` `fontScale` constant which contains the value `getContentSizeMultiplier` returned.
    - When the user changes the font scaling setting, fire `didUpdateDimensions`.
    - Don't expose the `scale` constant because we don't know how to get this value on iOS and we don't have an immediate need for it.

One question: do we need to make any changes to `UIManager`'s `scale` constant on Android? Is its name still suitable?

foghina commented:
That sounds like a breaking change on Android, so I'm not really in love with it. However, I did some grepping internally and I *think* we only have one place where we use this, so it should be easy to fix. I'd say let's go ahead with this plan.

Also, I just realized `scaledDensity` is exposed in two constants in UIManager: `windowPhysicalPixels.fontScale` and `screenPhysicalPixels.fontScale`. I'm guessing this is also iOS-inspired? Could we clean it up?

> Expose a UIManager fontScale constant which contains the value getContentSizeMultiplier returned.

Can you change constants on iOS? You can't on Android (without restarting the bridge). I'm asking because we should stick with the getter if we can't change the constant. We don't want to end up with `didUpdateDimensions` firing and the constant still having the old value.

Last question: what number does iOS return from `getContentSizeMultiplier`? Is it similar to `fontScale`? Could you use it in the same way on both platforms?

astreet commented:
>That sounds like a breaking change on Android, so I'm not really in love with it. However, I did some grepping internally and I think we only have one place where we use this, so it should be easy to fix. I'd say let's go ahead with this plan.

I only saw that one place too and it should really just be removed: we already factor system font scale on the native side by converting from SP so they're double-factoring font scale.

@rigdern changing `density` is a much riskier change than changing fontScale -- density is used a ton of places so I think we should just not touch that. 

Also, just expose `getResources().getConfiguration().fontScale` (instead of calculating it as you mentioned).

>On iOS: When the user changes the font scaling setting, fire didUpdateDimensions

I'm less sure about the impacts of this change... would it cause us to do a re-layout? Should probably get @javache in on this.

rigdern commented:
@foghina 

> Also, I just realized scaledDensity is exposed in two constants in UIManager: windowPhysicalPixels.fontScale and screenPhysicalPixels.fontScale. I'm guessing this is also iOS-inspired? Could we clean it up?

Why do you think this is iOS-inspired? Looking at the code, [iOS only exposes window dimensions](https://github.com/facebook/react-native/blob/b20b20656ac4b550300276a99bbb95d7365753ce/React/Modules/RCTUIManager.m#L1539-L1545) whereas [Android exposes both window and screen dimensions](https://github.com/facebook/react-native/blob/c144bbfb7ebdd174707638eaa344a9fddb38f253/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L103-L126). Do you have a suggestion for how this could be cleaned up?

> Can you change constants on iOS? You can't on Android (without restarting the bridge). I'm asking because we should stick with the getter if we can't change the constant. We don't want to end up with didUpdateDimensions firing and the constant still having the old value.

I don't think you can change constants on iOS but I don't think this is important in this case because it looks like apps don't use these constants directly. Instead, they are accessed through the `Dimensions` object which [gets updated whenever the dimensions change](https://github.com/facebook/react-native/blob/b20b20656ac4b550300276a99bbb95d7365753ce/Libraries/Utilities/Dimensions.js#L87-L89). We can update `Dimensions` whenever the `fontScale` changes as well.

> Last question: what number does iOS return from getContentSizeMultiplier? Is it similar to fontScale? Could you use it in the same way on both platforms?

If by `fontScale` you mean Android's `getResources().getConfiguration().fontScale` then, yes, empirically they seem to be the same. My team has been using `getResources().getConfiguration().fontScale` in our Android app and `getContentSizeMultiplier` in our iOS app. I do not understand where the values for `getContentSizeMultiplier` come from on iOS. [They are hard coded values](https://github.com/facebook/react-native/blob/b20b20656ac4b550300276a99bbb95d7365753ce/React/Modules/RCTAccessibilityManager.m#L138-L150) based on the [`preferredContentSizeCategory`](https://developer.apple.com/reference/uikit/uiapplication/1623048-preferredcontentsizecategory)

@astreet 

> @rigdern changing density is a much riskier change than changing fontScale -- density is used a ton of places so I think we should just not touch that.

Okay, I just noticed that [iOS already exports a `density` value under the name `scale`](https://github.com/facebook/react-native/blob/b20b20656ac4b550300276a99bbb95d7365753ce/React/Modules/RCTUIManager.m#L1543) so that's good.

>> On iOS: When the user changes the font scaling setting, fire didUpdateDimensions

> I'm less sure about the impacts of this change... would it cause us to do a re-layout? Should probably get @javache in on this.

Here are the options I have in mind for notifying the app that `fontScale` has changed:
  - Fire existing event `didUpdateDimensions`. Downside: this is an existing event. Apps might assume that when this event is fired, either `width` or `height` has changed and do something expensive in response. Firing it when `fontScale` has changed would break this assumption. (On a positive note, [I did a search for `didUpdateDimensions`](https://github.com/facebook/react-native/search?utf8=%E2%9C%93&q=didUpdateDimensions) and it appears the RN framework itself doesn't do anything expensive.)
  - Fire new event `didUpdateFontScale`.

Note that apps might access `fontScale` through `Dimensions.get('window').fontScale` and `PixelRation.getFontScale()`.

I think `didUpdateFontScale` is the better option. Reusing `didUpdateDimensions` was tempting because both dimensions and `fontScale` happen to be bundled under the `Dimensions` object. However, I think it makes sense that `didUpdateDimensions` fires for `width`/`height` changes and `didUpdateFontScale` fires for `fontScale` changes. Also, I'm not sure that it'd be common to want to handle both of these cases together which lessens the appeal of bundling them in a `didUpdateDimensions` event.

astreet commented:
I'm fine with creating a new event. Just making sure, updating the font scale causes us to perform a re-layout, right?

rigdern commented:
Sorry for the delay on this.

> Just making sure, updating the font scale causes us to perform a re-layout, right?

@astreet Yes. On iOS, text is marked as dirty and relaid out. On Android, the root view is destroyed and recreated.

I ran into some issues trying to implement this on Android:
1. **Updating JS `fontScale` value in response to system setting change.** After changing the system's font scale setting, the activity is [destroyed](https://github.com/facebook/react-native/blob/23e2610f4fcebbdad3ee8f0e18baa83c8bef624b/ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java#L136-L144) and [recreated](https://github.com/facebook/react-native/blob/23e2610f4fcebbdad3ee8f0e18baa83c8bef624b/ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java#L86-L97). However, this does not result in the native constants being rexported so `Dimensions.get('window').fontScale` still refers to the initial value. I can fire the JS event `didUpdateFontScale` to update JS however I don't know what native event `UIManager` should listen to for this. Any ideas?
Also, we'll probably end up 2 doing big rerenders. One for recreating the root view and one for JS code in the app that is listening to `didUpdateFontScale`. I'm not sure how to avoid 2 rerenders.

1. **Exposing `getResources().getConfiguration().fontScale`.** @astreet, you suggested exposing `fontScale` as `getResources().getConfiguration().fontScale` rather than calculating it from `DisplayMetrics.scaledDensity / DisplayMetrics.density`. However, there are 2 different `fontScales` exposed: one for [`windowPhysicalPixels`](https://github.com/facebook/react-native/blob/23e2610f4fcebbdad3ee8f0e18baa83c8bef624b/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L111-L112) and one for [`screenPhysicalPixels`](https://github.com/facebook/react-native/blob/23e2610f4fcebbdad3ee8f0e18baa83c8bef624b/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java#L123-L124). Which one does `getResources().getConfiguration().fontScale` refer to and how do I calculate the other one?

1. **window vs screen display metrics.** The RN Java code says that [`getWindowDisplayMetrics` is deprecated](https://github.com/facebook/react-native/blob/23e2610f4fcebbdad3ee8f0e18baa83c8bef624b/ReactAndroid/src/main/java/com/facebook/react/uimanager/DisplayMetricsHolder.java#L91-L98) but `PixelRatio.getFontScale()` uses the window display metrics (`Dimensions.get('window')`) rather than the screen display metrics (`Dimensions.get('screen')`). Even more confusing, I think iOS doesn't even expose `Dimensions.get('screen')` which is supposedly recommended over the deprecated `Dimensions.get('window')`.

rigdern commented:
@astreet do you have any thoughts on [my previous comment](https://github.com/facebook/react-native/pull/11008#issuecomment-267463468)?

rigdern commented:
@astreet 

Thanks, I will try using `onConfigurationChanged`.

Regarding `window` vs `screen`, you answered my question. I wasn't aware of the difference between `window` and `screen` and now I know there is no difference as far as font scale goes.

I'm confused about the deprecation of `window`. The comment says `window` is deprecated but iOS only provides `window`. Should iOS start providing `screen` and then we can stop referencing `window` within the RN codebase?

andreicoman11 commented:
Re: `window` vs `screen`.

`window` is indeed deprecated and we should be using `screen` everywhere. 
On iOS `screen` and `window` are the same. Native provides only `window`, but we actually expose that data as both `window` and `screen` because of this: https://github.com/facebook/react-native/blob/master/Libraries/Utilities/Dimensions.js#L56. Ideally, we should be providing everything directly as `screen` and removing `window`. Basically, what you mentioned.
On Android, `window` and `screen` are different, but both are provided. Eventually we should be not providing `window` at all and using `screen` everywhere.

rigdern commented:
@andreicoman11 I see, thanks for the explanation. I suspect the [code you linked to](https://github.com/facebook/react-native/blob/0623a47ecd3cc7546d6a5581b433388898b017cb/Libraries/Utilities/Dimensions.js#L56) is not running on iOS. The reason is that the code is within [an if statement checking for `dim.windowPhysicalPixels`](https://github.com/facebook/react-native/blob/0623a47ecd3cc7546d6a5581b433388898b017cb/Libraries/Utilities/Dimensions.js#L32). However, iOS doesn't export its dimensions under the property `windowPhysicalPixels`. Instead, [iOS exports them as `window`](https://github.com/facebook/react-native/blob/f839b58c200ce25abfcbc8b2c57b3012f4de14c3/React/Modules/RCTUIManager.m#L1539-L1545).

So based on code inspection, I don't think iOS currently supports `screen` at all.

rigdern commented:
@astreet 

>> Updating JS fontScale value in response to system setting change.

> Ah, that's a good point, you can listen from onConfigurationChanged.

I looked into using `onConfigurationChanged` but it looks like that requires `android:configChanges="fontScale"`. I have a couple of concerns about that:
  1. This would require React Native apps to set `android:configChanges="fontScale"` which is not currently the default for new RN apps.
  2. Earlier, @foghina mentioned concerns that `android:configChanges="fontScale"` is likely untested and there will probably be bugs. I'm worried about the potential regressions such a change could cause.

Instead of supporting `android:configChanges="fontScale"`, I tried calling into the `UIManager` when the activity is created to give the` UIManager` the opportunity to fire an event in the case of font scale changes. I'm not familiar enough with the Android or RN lifecycle to know if this is on the right track. Here's what that looks like: https://github.com/rigdern/react-native/commit/62952dbf076081f03a0118f0f56f7354ca7ff834#diff-f3ee27db155d4af4368a3ff1e60a765c

Do you have any thoughts on this?

astreet commented:
@foghina correct me if I'm wrong but I think he meant that changing the font scale in RN by any means might cause regressions (as you need to re-layout). I actually think you're approach is easier however: if we set `android:configChanges="fontScale"` we will have to manually retrigger a layout, and it could also break other things in apps if the rest of their non-RN app doesn't also manually trigger a re-layout.

The diff you posted looks like the right approach, I'd just say we should add a `onCreate` lifecycle event and listen to that from UIManager instead of putting it in the delegate which is more fragile.

rigdern commented:
@astreet Thanks for the information. I pushed a commit which assumes that `android:configChanges="fontScale"` is **not** set. It leverages the `onHostResume` hook which `UIManager` is already using.

My commit also implements the rest of the ideas we discussed here. [My PR description](https://github.com/facebook/react-native/pull/11008#issue-190214018) was updated to accurately summarize the current approach.

I created a corresponding PR (#12268) for iOS. There is one open issue. When `didUpdateDimensions` fires, the event contains different property names on iOS and Android.
  - On Android, the root property names are `windowPhysicalPixels` and `screenPhysicalPixels`.
  - On iOS, the root property name is `window`.

We should expose the same property name on both platforms. What should it be?

rigdern commented:
@astreet I added a dependency to `UIManagerModule` on the `DeviceEventManagerModule` manager so that `UIManager` can fire events. However, it lead to this error from buck:

```
BUILD FAILED: Cycle found:
//ReactAndroid/src/main/java/com/facebook/react/devsupport:devsupport ->
//ReactAndroid/src/main/java/com/facebook/react/modules/debug:debug ->
//ReactAndroid/src/main/java/com/facebook/react/uimanager:uimanager ->
//ReactAndroid/src/main/java/com/facebook/react/modules/core:core ->
//ReactAndroid/src/main/java/com/facebook/react/devsupport:devsupport
```

Any ideas on how to break the dependency cycle?

astreet commented:
>We should expose the same property name on both platforms. What should it be?

We should expose it the same as they are exposed as initial constants from UIManager: we wrap those in Dimensions, so maybe Dimensions should handle exposing the callback to JS code. I agree that we should figure out the naming, I really don't have an opinion -- do what you think is best and I'll submit it for review internally.

astreet commented:
>Any ideas on how to break the dependency cycle?

Imo, seems like our core modules shouldn't be depending on devsupport, probably by providing an interface for the DevManager and exposing it in a separate buck module (devsupport:interfaces) that ExceptionsManagerModule can take a dependency on instead.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java:
```diff
@@ -539,6 +551,37 @@ public void sendAccessibilityEvent(int tag, int eventType) {
     mUIImplementation.sendAccessibilityEvent(tag, eventType);
   }
 
+  public void emitUpdateDimensionsEvent() {
```
I know you're not adding this, but seems like we should be sharing code with where we create this map when exposing the initial constants.

rigdern replied:
I fixed this now. One interesting detail is that the dimensions part of constants is now a `WritableMap` instead of a `HashMap`. I wasn't sure what the impact of this would be. It built and ran fine.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModuleConstants.java:
```diff
@@ -81,7 +81,7 @@
   }
 
-  public static Map<String, Object> getConstants() {
+  public static Map<String, Object> getConstants(float fontScale) {
```
Didn't we decide we can get this off of displayMetrics as scaledDensity / density? At the least, I'd rather just pass in the Configuration object if you're worried about loss of precision

rigdern replied:
In [this earlier comment](https://github.com/facebook/react-native/pull/11008#issuecomment-263545377) you mentioned that we should get the font scale from `getResources().getConfiguration().fontScale` rather than calculating it from `scaledDensity / density`.

The reason I pass `fontScale` around rather than the Configuration object is that I wanted it to be very obvious that we are passing `mFontScale` to JavaScript. I was concerned that there might be some race where font scale is extracted from the Configuration object but it has a different value than `mFontScale`.

astreet replied:
Ok, this sounds reasonable

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java:
```diff
@@ -539,6 +551,37 @@ public void sendAccessibilityEvent(int tag, int eventType) {
   }
 
+  public void emitUpdateDimensionsEvent() {
+    mFontScale = getReactApplicationContext().getResources().getConfiguration().fontScale;
```
imo, this shouldn't be in emitUpdate() given the name. I'd just set it in onHostResume

rigdern commented:
@astreet 

I started working on breaking the dependency cycle: bb214ef. Is that along the lines you were thinking? Here are some concerns regarding that commit:
  - Is it normal to have BUCK files at non-leaf directories? The devsupport interfaces are in a subdirectory of `devsupport`. Both `devsupport` and `devsupport/interfaces` contain BUCK files.
  - I changed the namespace of the devsupport interfaces from `com.facebook.react.devsupport` to `com.facebook.react.devsupport.interfaces`. A couple of questions regarding this:
    - Is that an okay namespace name?
    - Do app developers use devsupport in their apps/modules? If so, this would be a breaking change.
  - There are some interfaces that are still in `devsupport` that I didn't need to move to `devsupport/interfaces` in order to break the cycle. Should I move those other interfaces to `devsupport/interfaces` anyway?
  - `PackagerStatusCallback` is strange because it used to be inside of a class (`DevServerHelper`) and now it lives standalone outside of any class.

Now I have a new dependency cycle:

```
BUILD FAILED: Cycle found:
//ReactAndroid/src/main/java/com/facebook/react/uimanager:uimanager ->
//ReactAndroid/src/main/java/com/facebook/react/modules/core:core ->
//ReactAndroid/src/main/java/com/facebook/react/jstasks:jstasks ->
//ReactAndroid/src/main/java/com/facebook/react/uimanager:uimanager
```

How should I approach breaking this cycle? I don't have enough of an understanding of the RN Android architecture to know which dependency is inappropriate 😦 . Maybe there needs to be a `UIManager` interface? If so, what should it be called?

> > We should expose the same property name on both platforms. What should it be?

> We should expose it the same as they are exposed as initial constants from UIManager: we wrap those in Dimensions, so maybe Dimensions should handle exposing the callback to JS code. I agree that we should figure out the naming, I really don't have an opinion -- do what you think is best and I'll submit it for review internally.

My plan is to expose an event called `change` off of `Dimensions` (similar to the event API that `AppState` offers). The event object will contain the data under the property `screen`.

astreet commented:
>I started working on breaking the dependency cycle: bb214ef. Is that along the lines you were thinking? Here are some concerns regarding that commit

Yep, that's basically what I'm thinking. What we more generally do though is use one buck file and define a 'regular' rule and an 'interfaces' rule from it in the same buck file. In this way, you'd be able to depend on 'devsupport:interfaces' instead of 'devsupport/interfaces:interfaces'. Note that you can still have your `interfaces` directory and package, it just wouldn't have its own buck file.

As far as it being breaking, yes, this might be breaking. AFAIK, this is not a common part of our core API so I'm ok with it, just title the commit that it's BREAKING so it goes in the breaking changes doc.

>Another dep cycle

Ugh, sorry. IMO, jstasks shouldn't depend on UIManager: it doesn't use it at all. Looking at the code, it looks like it's a dependency only because of AppRegistry being in uimanager which doesn't make sense to me. I would be fine moving AppRegistry to its own buck module (`react/appregistry:appregistry`).

>My plan is to expose an event called change off of Dimensions (similar to the event API that AppState offers). The event object will contain the data under the property screen.

sgtm

astreet commented:
I'd also suggest doing the buck changes as separate PRs: prevents problems if something needs to be reverted (e.g. we won't lose the whole PR), makes things clearer in the history, and will go ahead and get parts landed and out of mind :)

rigdern commented:
@astreet I did the buck changes in PR #12329. Please take a look at it. Once we get the dependency cycle stuff resolved and checked in, I'll resume this PR.

rigdern commented:
@astreet I did all of the changes we discussed. I rebased this on top of my commit for breaking dependency cycles and I added a `change` event to `Dimensions`.

astreet commented:
Can we add tests for this? This is such an infrequent event that I'm worried we might break it accidentally and not realize it.

rigdern commented:
@astreet Tests for which event? `didUpdateDimensions`? The `change` event for the `Dimensions` object?

Normally, firing these events requires rotating the device or changing the font scale settings. How do you propose to test it? For `change`, I could call `Dimensions.set()` and very that `change` fires.

astreet commented:
This would be a test to make sure that when we detect a font scale change, that JS receives an update and triggers listeners. We have instrumentation tests that test JS and native code together here: https://github.com/facebook/react-native/tree/master/ReactAndroid/src/androidTest/java/com/facebook/react/tests.

Simulating a configuration such that it changes between onResume calls might be difficult: since it's running as an actual Application on an actual device (so that we can test JS), we can't do standard mocking of static functions with PowerMockito afaik. An idea that comes to mind is creating a setter function on UIManager, but it's not great. Let me know if you have any ideas. Really sketchy other idea: fontScale on Configuration isn't final so you might be able to just manually set it...
